drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,869 @@
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	func_clear_dme[id](dma);
+	func_clear_fe[id](dma);
+#else
+	func_clear_gi[id](dma);
+#endif
+}
+
+static void dma_stm32_irq_handler(void *arg)
+{
+	struct device *dev = arg;
+	struct dma_stm32_data *data = dev->driver_data;
+	const struct dma_stm32_config *config = dev->config->config_info;
+	DMA_TypeDef *dma = (DMA_TypeDef *)(config->base);
+	struct dma_stm32_stream *stream;
+	int id;
+
+	for (id = 0; id < data->max_streams; id++) {
+		if (func_is_active_tc[id](dma) ||
+		    func_is_active_ht[id](dma) ||
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+		    func_is_active_fe[id](dma) ||
+#endif
+		    func_is_active_te[id](dma)) {
+			break;
+		}
+	}
+
+	if (id == data->max_streams) {
+		LOG_ERR("Unknow interrupt happened.");
+	}
+
+	stream = &data->streams[id];
+	stream->busy = false;
+	if (func_is_active_tc[id](dma)) {
+		func_clear_tc[id](dma);
+
+		stream->dma_callback(stream->callback_arg, id, 0);
+	} else if (func_is_active_ht[id](dma)) {
+		/* Silently ignore spurious transfer half complete IRQ */
+		LOG_INF("Half of the transfer has been completed.");
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+	} else if (func_is_active_fe[id](dma)) {
+		LOG_ERR("FiFo error.");
+		dma_stm32_dump_stream_irq(dev, id);
+		dma_stm32_clear_stream_irq(dev, id);
+
+		stream->dma_callback(stream->callback_arg, id, -EIO);
+#endif
+	} else {
+		LOG_ERR("Transfer Error.");
+		dma_stm32_dump_stream_irq(dev, id);
+		dma_stm32_clear_stream_irq(dev, id);
+
+		stream->dma_callback(stream->callback_arg, id, -EIO);
+	}
+}
+
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+static void dma_stm32_config_direct_mode(DMA_TypeDef *dma, u32_t id, bool on)
+{
+	if (on) {
+		LL_DMA_EnableIT_DME(dma, table_stream[id]);
+		LL_DMA_DisableIT_FE(dma, table_stream[id]);
+		LL_DMA_DisableFifoMode(dma, table_stream[id]);
+	} else {
+		LL_DMA_DisableIT_DME(dma, table_stream[id]);
+		LL_DMA_EnableIT_FE(dma, table_stream[id]);
+		LL_DMA_EnableFifoMode(dma, table_stream[id]);
+	}
+}
+#endif
+
+#if defined(CONFIG_SOC_SERIES_STM32F2X) || \
+	defined(CONFIG_SOC_SERIES_STM32F4X) || \
+	defined(CONFIG_SOC_SERIES_STM32F7X)
+#define BURST_SIZE(dev, dev_short)					\
+do {									\
+	case 1:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_SINGLE);	\
+		break;							\
+	case 4:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC4);	\
+		break;							\
+	case 8:								\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC8);	\
+		break;							\
+	case 16:							\
+		LL_DMA_Set##dev##Burstxfer(dma, table_stream[id],	\
+				LL_DMA_##dev_short##BURST_INC16);	\
+		break;							\
+	default:							\
```

Comment:
the cases here should be 0,1,2,3 and not 1,4,8,16

Reply:
This value represents how many data units does one burst send/receive, I don't understand why it should be 0, 1, 2, 3, comment in dma.h is blur about this. After searching 'source_burst_length' in Zephyr's tree, I see some usage in dma_qmsi.c, dma_sam_xdmac.c, dma_cavs.c, they treat this value with find_msb_set and something like that. This value should be 0 or 2^n I guess, right?

Reply:
It seems again a mismatch like the data size. Before the driver was using dma_burst_index(), 
which was expecting the burst to be the exponent (i.e. 0 is 8-bit burst, 1 is 16-bit burst, ...).

But all the drivers seems to expect the burst size like 1, 4, 8, 16. So, we may change I2S driver
according to the API (even if in the burst case the dma.h seems not really clear in the 
explanation).